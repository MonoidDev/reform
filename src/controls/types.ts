import { Either } from '../types/Either';
import { ErrorMessage } from '../types/ErrorMessage';
import type { FormControl } from './FormControl'
import type { StructFormControl } from './StructFormControl'

export type AnyFormControl = FormControl<any, any, any, any> | StructFormControl<any>;

export type FormResult<O> = Either<ErrorMessage | undefined, O>;

export type Unsubscribe = () => void;
