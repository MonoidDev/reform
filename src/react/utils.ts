import { useState } from 'react';

export const useForceUpdate = () => {
  const [, s] = useState({});
  return () => s({});
}
