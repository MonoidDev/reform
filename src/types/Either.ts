export type Left<T> = {
  _tag: 'left';
  left: T;
};

export type Right<T> = {
  _tag: 'right';
  right: T;
};

export type Either<L, R> = Left<L> | Right<R>;

export function isLeft<L, R>(e: Either<L, R>): e is Left<L> {
  return e._tag === 'left';
}

export function isRight<L, R>(e: Either<L, R>): e is Right<R> {
  return e._tag === 'right';
}

export function makeLeft<T>(left: T): Left<T> {
  return {
    _tag: 'left',
    left,
  };
}

export function makeRight<T>(right: T): Right<T> {
  return {
    _tag: 'right',
    right,
  };
}
