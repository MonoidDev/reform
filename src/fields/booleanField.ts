import { any } from '../types/BasicTypes';
import { Either, makeLeft, makeRight } from '../types/Either';
import { ErrorMessage } from '../types/ErrorMessage';

export const booleanField = (options: BooleanFieldOptions = {}) => {
  const { message, truthy = 'true', required } = options;

  const booleanType = any().refine((input) => {
    if (typeof input === 'boolean') return makeRight(input);
    if (input === truthy) return makeRight(true);
    if (input === '')
      return required
        ? makeLeft({ message: message ?? '' })
        : makeRight(undefined);
    return makeRight(false);
  });

  return {
    ...booleanType,
    refine<O2, E2 extends ErrorMessage = ErrorMessage>(
      validate: (input: boolean | undefined) => Either<E2, O2>,
    ) {
      return {
        ...this,
        ...booleanType.refine(validate),
      };
    },
    map<O2>(mapper: (input: boolean | undefined) => O2) {
      return {
        ...this,
        ...booleanType.map(mapper),
      };
    },
    true(message: string) {
      return this.refine((input) =>
        input === true ? makeRight(input) : makeLeft({ message }),
      );
    },
    false(message: string) {
      return this.refine((input) =>
        input === false ? makeRight(input) : makeLeft({ message }),
      );
    },
  };
};

export interface BooleanFieldOptions {
  message?: string;
  truthy?: any; // Equal to what value when we consider it truthy, by default 'true'.
  required?: boolean;
}
