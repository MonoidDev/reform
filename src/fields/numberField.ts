import { any } from '../types/BasicTypes';
import { Either, makeLeft, makeRight } from '../types/Either';
import { ErrorMessage } from '../types/ErrorMessage';
import { Resolver } from '../types/Resolver';

export interface NumberFieldOptions {
  message: string;
}

export const numberField = (
  options: NumberFieldOptions
) => {
  const { message } = options;

  const numberType = any().refine((input) => {
    if (
      typeof input === 'string' &&
      input.trim().length === 0
    ) {
      return makeRight(undefined);
    }

    const parsed = Number.parseFloat(input);
    if (Number.isNaN(parsed)) {
      return makeLeft({ message });
    }
    return makeRight(parsed);
  });

  return makeNumberField(numberType);
};

export const requiredNumberField = (options: RequiredNumberFieldOptions) => {
  const { message, requiredMessage } = options;

  const numberType = any().refine((input) => {
    const parsed = Number.parseFloat(input);
    if (typeof input === 'string' && input.trim() === '') {
      return makeLeft({ message: requiredMessage });
    }

    if (Number.isNaN(parsed)) {
      return makeLeft({ message });
    }
    return makeRight(parsed);
  });

  return makeNumberField(numberType);
}

export interface RequiredNumberFieldOptions extends NumberFieldOptions {
  requiredMessage: string;
}

const makeNumberField = <TNumber extends number | undefined>(numberType: Resolver<any, TNumber, ErrorMessage, string>) => {
  return {
    ...numberType,
    refine<O2, E2 extends ErrorMessage = ErrorMessage>(
      validate: (input: TNumber) => Either<E2, O2>
    ) {
      return {
        ...this,
        ...numberType.refine(validate),
      };
    },
    map<O2>(mapper: (input: TNumber) => O2) {
      return {
        ...this,
        ...numberType.map(mapper),
      };
    },
    max(n: number, message: string) {
      return this.refine((input) =>
        input === undefined || input > n
          ? makeLeft({ message })
          : makeRight(input)
      );
    },
    min(n: number, message: string) {
      return this.refine((input) =>
        input === undefined || input < n
          ? makeLeft({ message })
          : makeRight(input)
      );
    },
    positive(message: string) {
      return this.refine((input) =>
        input === undefined || input <= 0
          ? makeLeft({ message })
          : makeRight(input)
      );
    },
    negative(message: string) {
      return this.refine((input) =>
        input === undefined || input >= 0
          ? makeLeft({ message })
          : makeRight(input)
      );
    },
    integer(message: string) {
      return this.refine((input) =>
        !Number.isSafeInteger(input) ? makeLeft({ message }) : makeRight(input)
      );
    },
  };
}