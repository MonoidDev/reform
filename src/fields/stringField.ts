import { string } from '../types/BasicTypes';
import { Either, makeLeft, makeRight } from '../types/Either';
import { ErrorMessage } from '../types/ErrorMessage';

export const stringField = () => {
  const stringType = string<string>();

  return {
    ...stringType,
    refine<O2, E2 extends ErrorMessage = ErrorMessage>(
      validate: (input: string) => Either<E2, O2>
    ) {
      return {
        ...this,
        ...stringType.refine(validate),
      };
    },
    map<O2>(mapper: (input: string) => O2) {
      return {
        ...this,
        ...stringType.map(mapper),
      };
    },
    trim() {
      return this.map((s) => s.trim());
    },
    matches(regexp: RegExp, message: string) {
      return this.refine((input) =>
        regexp.test(input) ? makeRight(input) : makeLeft({ message })
      );
    },
    max(n: number, message: string) {
      return this.refine((input) => input.length > n ? makeLeft({ message }) : makeRight(input));
    },
    min(n: number, message: string) {
      return this.refine((input) => input.length < n ? makeLeft({ message }) : makeRight(input));
    },
    required(message: string) {
      return this.trim().min(1, message);
    },
  };
};
