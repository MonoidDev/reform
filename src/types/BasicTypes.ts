import { isLeft, makeLeft, makeRight } from './Either';
import { ErrorMessage } from './ErrorMessage';
import { AnyResolver, OutputOf } from './infer';
import {
  ErrorContext,
  makeErrorContext,
  makeResolver,
  Resolver,
} from './Resolver';
import { isObject } from './utils';

export interface TAny extends Resolver<any, any, ErrorMessage, 'any'> {}

export function any(): TAny {
  return {
    ...makeResolver<any, any, ErrorMessage, 'any'>('any', (input) =>
      makeRight(input)
    ),
  };
}

export interface TString
  extends Resolver<unknown, string, ErrorMessage, 'string'> {}

export function string(): TString {
  return {
    ...makeResolver<unknown, string, ErrorMessage, 'string'>(
      'string',
      (input) => {
        if (typeof input === 'string') {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect a string, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TNumber
  extends Resolver<unknown, number, ErrorMessage, 'number'> {}

export function number(): TNumber {
  return {
    ...makeResolver<unknown, number, ErrorMessage, 'number'>(
      'number',
      (input) => {
        if (typeof input === 'number') {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect a number, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TBoolean
  extends Resolver<unknown, boolean, ErrorMessage, 'boolean'> {}

export function boolean(): TBoolean {
  return {
    ...makeResolver<unknown, boolean, ErrorMessage, 'boolean'>(
      'boolean',
      (input) => {
        if (typeof input === 'boolean') {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect a boolean, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TNull extends Resolver<unknown, null, ErrorMessage, 'null'> {}

export function nullType(): TNull {
  return {
    ...makeResolver<unknown, null, ErrorMessage, 'null'>('null', (input) => {
      if (input === null) {
        return makeRight(input);
      } else {
        return makeLeft({
          message: `Expect a null, got ${typeof input}. `,
        });
      }
    }),
  };
}

export interface TUndefined
  extends Resolver<unknown, undefined, ErrorMessage, 'undefined'> {}

export function undefinedType(): TUndefined {
  return {
    ...makeResolver<unknown, undefined, ErrorMessage, 'undefined'>(
      'undefined',
      (input) => {
        if (input === undefined) {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect an undefined, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TObject
  extends Resolver<unknown, {}, ErrorMessage, 'object'> {}

export function object(): TObject {
  return {
    ...makeResolver<unknown, {}, ErrorMessage, 'object'>('object', (input) => {
      if (isObject(input)) {
        return makeRight(input);
      } else {
        return makeLeft({
          message: `Expect an object, got ${typeof input}. `,
        });
      }
    }),
  };
}

export interface TArray<T extends AnyResolver>
  extends Resolver<unknown, OutputOf<T>[], ErrorMessage, 'array'> {}

export function array<T extends AnyResolver>(t: T): TArray<T> {
  return {
    ...makeResolver<unknown, OutputOf<T>[], ErrorMessage, 'array'>('array'),
    resolve(input, contexts = [makeErrorContext()]) {
      if (!Array.isArray(input)) {
        contexts[contexts.length - 1].error = {
          message: `Expect an array, got ${typeof input}`,
        };
        return makeLeft(contexts);
      }

      let value: any = [];

      const newErrorContexts: ErrorContext[] = []; 
      for (let i = 0; i < input.length; i++) {
        const nextContext: ErrorContext = {
          inputPath: [...contexts[contexts.length - 1].inputPath, i],
          resolverPath: [...contexts[contexts.length - 1].resolverPath],
          error: undefined,
        };

        const result = t.resolve(input[i], [nextContext]);

        if (isLeft(result)) {
          newErrorContexts.push(...result.left);
        } else {
          value.push(result.right);
        }
      }

      if (newErrorContexts.length) {
        contexts.pop(); // Remove last context because there're no errors
        return makeLeft([
          ...contexts,
          ...newErrorContexts,
        ]);
      }

      return makeRight(value);
    }
  };
}
