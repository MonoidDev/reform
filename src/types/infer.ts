import { TAny } from "./BasicTypes"
import { Resolver } from "./Resolver"

export type AnyResolver = Resolver<any, any, any, string>;

export type ResolverMap = {
  [K in string]: AnyResolver;
}

export type InputOfResolverMap<RM extends ResolverMap> = {
  [K in keyof RM]: InputOf<RM[K]>;
}

export type OutputOfResolverMap<RM extends ResolverMap> = {
  [K in keyof RM]: OutputOf<RM[K]>;
}

export type InputOf<T extends AnyResolver> = T extends Resolver<infer I, any, any, any> ? I : never;

export type OutputOf<T extends AnyResolver> = T extends Resolver<any, infer O, any, any> ? O : never;
