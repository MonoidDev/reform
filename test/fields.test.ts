import { booleanField } from '../src/fields/booleanField';
import { numberField, requiredNumberField } from '../src/fields/numberField';
import { stringField } from '../src/fields/stringField';
import { string } from '../src/types/BasicTypes';
import { makeLeft, makeRight } from '../src/types/Either';
import { OutputOf } from '../src/types/infer';
import { makeResolver } from '../src/types/Resolver';
import { expectLeft, expectRight } from './utils';

describe('string', () => {
  it('can pipe', () => {
    const requiredString = string().pipe(
      makeResolver('requiredString', (input) => {
        if (input.length > 0) {
          return makeRight(input);
        }
        return makeLeft({
          message: 'required',
        });
      })
    );

    expectLeft(requiredString.resolve(''));
    expect(requiredString.resolve('')).toMatchSnapshot();
    expectRight(requiredString.resolve('xxx'));
  });

  it('can refine', () => {
    const requiredString = string().refine((input) =>
      input.length > 0 ? makeRight(input) : makeLeft({ message: 'required' })
    );

    expectLeft(requiredString.resolve(''));
    expect(requiredString.resolve('')).toMatchSnapshot();
    expectRight(requiredString.resolve('xxx'));
  });
});

describe('stringField', () => {
  it('can map', () => {
    expectRight(
      stringField()
        .map((s) => s.length)
        .resolve('123')
    ).toBe(3);
    expectRight(stringField().trim().resolve('     ')).toBe('');
  });

  it('can be required', () => {
    expectLeft(stringField().required('x').resolve('    '));
    expectRight(stringField().required('x').resolve(' xxx   ')).toBe('xxx');
  });
});

describe('numberField', () => {
  it('can type', () => {
    const nullable = numberField({ message: 'x' });
    const nullableV1: OutputOf<typeof nullable> = 11;
    const nullableV2: OutputOf<typeof nullable> = undefined;

    const required = requiredNumberField({ message: 'x', requiredMessage: 'x' });
    const requiredV1: OutputOf<typeof required> = 11;
    //@ts-expect-error
    const requiredV2: OutputOf<typeof required> = undefined;
  });

  it('can validate', () => {
    expectLeft(numberField({ message: 'oops' }).resolve('eqwe'));
    expectRight(numberField({ message: 'sda' }).resolve('213'));
    expectRight(numberField({ message: 'sda' }).min(3, '3').resolve(12312));
    expectLeft(numberField({ message: 'sda' }).min(3, '3').resolve(2));
    expectRight(numberField({ message: 'sda' }).integer('i').resolve(2));
    expectLeft(numberField({ message: 'sda' }).integer('i').resolve(0.0001));
    expectRight(numberField({ message: 's' }).resolve('')).toBe(undefined);
    expectLeft(requiredNumberField({ message: 's', requiredMessage: 's' }).resolve(''));
  });
});

describe('booleanField', () => {
  it('can validate', () => {
    expectRight(
      booleanField({ message: 'oops', required: true }).resolve('true')
    );
    expectRight(booleanField({ message: 'oops' }).resolve('')).toBe(undefined);
  });
});
