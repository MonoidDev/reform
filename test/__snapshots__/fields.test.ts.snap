// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`string can pipe 1`] = `
Object {
  "_tag": "left",
  "left": Array [
    Object {
      "error": Object {
        "message": "required",
      },
      "inputPath": Array [],
      "resolverPath": Array [],
    },
  ],
}
`;

exports[`string can refine 1`] = `
Object {
  "_tag": "left",
  "left": Array [
    Object {
      "error": Object {
        "message": "required",
      },
      "inputPath": Array [],
      "resolverPath": Array [],
    },
  ],
}
`;
